File: app/src/main/java/com/fonteraro/martaco/cinemahub/ui/details/DetailsViewModel.java
package com.fonteraro.martaco.cinemahub.ui.details;

import android.arch.lifecycle.LiveData;
import android.support.annotation.NonNull;

import com.fonteraro.martaco.cinemahub.data.DataRepository;
import com.fonteraro.martaco.cinemahub.data.model.db.MovieDetails;
import com.fonteraro.martaco.cinemahub.ui.base.BaseViewModel;
import com.fonteraro.martaco.cinemahub.ui.details.callback.MovieRequestListener;

public class DetailsViewModel extends BaseViewModel {

    private LiveData<MovieDetails> movieDetails;

    public DetailsViewModel(DataRepository dataRepository) {
        super(dataRepository);
    }

    public void getMovieDetails(@NonNull String id, @NonNull MovieRequestListener listener) {
        getDataRepository().getMovie(id, listener);
    }

    public void setMovieDetails(LiveData<MovieDetails> movieDetails) {
        this.movieDetails = movieDetails;

    }

    public MovieDetails getMovieDetails() {
        if (movieDetails != null) {
            return movieDetails.getValue();
        }

        return null;
    }
}

File: app/src/main/java/com/fonteraro/martaco/cinemahub/ui/details/DetailsViewModel.kt
package com.fonteraro.martaco.cinemahub.ui.details

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.fonteraro.martaco.cinemahub.ui.base.BaseViewModel
import com.fonteraro.martaco.cinemahub.data.MoviesRepository
import com.fonteraro.martaco.cinemahub.di.module.CoroutineDispatchers
import com.fonteraro.martaco.cinemahub.ui.model.Movie
import com.fonteraro.martaco.cinemahub.utils.Event
import com.fonteraro.martaco.cinemahub.utils.extensions.toEvent
import kotlinx.coroutines.launch
import javax.inject.Inject

class DetailsViewModel @Inject constructor(
    private val moviesRepository: MoviesRepository,
    dispatchers: CoroutineDispatchers
) : BaseViewModel(dispatchers = dispatchers) {

    private var _movie = MutableLiveData<Event<Movie>>()

    fun getMovieDetails(movieId: Long) = ioScope.launch {
        moviesRepository.getMovie(movieId).also {
            uiScope.launch { _movie.value = it.toEvent() }
        }
    }

    val movie: LiveData<Event<Movie>>
    get() = _movie
}
